rpc/src/transaction_status_service.rs

```diff
@@ -34,47 +36,63 @@ impl TransactionStatusService {
         enable_extended_tx_metadata_storage: bool,
         exit: Arc<AtomicBool>,
     ) -> Self {
-        let thread_hdl = Builder::new()
-            .name("solTxStatusWrtr".to_string())
-            .spawn(move || {
-                info!("TransactionStatusService has started");
-                loop {
-                    if exit.load(Ordering::Relaxed) {
-                        break;
-                    }
-
-                    let message = match write_transaction_status_receiver
-                        .recv_timeout(Duration::from_secs(1))
-                    {
-                        Ok(message) => message,
-                        Err(RecvTimeoutError::Disconnected) => {
-                            break;
+        let mut thread_handles = vec![];
+        let write_transaction_status_receiver =
+            Arc::new(Mutex::new(write_transaction_status_receiver));
+
+        for worker in 0..NUM_TSS_WORKER_THREADS {
+            let write_transaction_status_receiver = Arc::clone(&write_transaction_status_receiver);
+            let max_complete_transaction_status_slot =
+                Arc::clone(&max_complete_transaction_status_slot);
+            let transaction_notifier = transaction_notifier.clone();
+            let blockstore = Arc::clone(&blockstore);
+            let exit = Arc::clone(&exit);
+
+            let thread_hdl = Builder::new()
+                .name(format!("solTxStatusWrtr-{}", worker))
+                .spawn(move || {
+                    info!("TransactionStatusService worker {worker} has started");
+                    loop {
+                        if exit.load(Ordering::Relaxed) {
+                                break;
+                            }
+
+                        let message = {
+                            let tss_receiver = write_transaction_status_receiver.lock().unwrap();
+                            match tss_receiver.recv_timeout(Duration::from_secs(1)) {
```

Comment:
I agree with your second comment, we have the opposite problem here. On master, we are behind and have 80k pending messages. So we would never sleep. With these changes and all the gains, I am still seeing 500-1000 pending msg. So It still won't sleep. So trying to tune a sleep would be moving in the wrong direction here.

Reply:
@fkouteib is your comment true for master + mainnet load? Or are you only referring to master + high TPS synthetic load?

Reply:
Sorry I should have been more specific. Here, I mean master branch code with high TPS in my experiment setup. And same applies for numbers quoted for improvements

Reply:
I have not profiled anything for a mainnet workload. And I don't think we have these channel sizes in any metrics afaik.

Reply:
@steviez I don't understand your high level concern of how the polling negatively affects anything for this service itself being able to keep up, or just its normal operation. If it's falling behind, it's never polling. If it's ever polling and empty for 1 sec, thats almost 3 slots with no tx. So the network is down (?) since we should be getting vote tx at minimum. A sleep on the other hand is a manual knob for this service that would have to be tuned every so often as we increase TPS. If the sleep optimization is about increasing CPU availability for other threads on other services, then it may be worth looking into but seems like a separate investigation and out of scope here.

Reply:
> If it's falling behind, it's never polling.

Correct, if the channel is never empty, then we'll hit [this case](https://github.com/crossbeam-rs/crossbeam/blob/45425b032b75d40c8f79be2133eb7d33aaa1d4e4/crossbeam-channel/src/flavors/list.rs#L463-L467) in `recv_timeout()` and return immediately.

> If it's ever polling and empty for 1 sec, thats almost 3 slots with no tx. So the network is down (?) since we should be getting vote tx at minimum

There are still some known "glitches" so the network isn't necessarily down in this case. For example, while better with the activation of partitioned epoch rewards, epoch boundary slots still take a while. Looking at my test node, it spent ~1.7s creating the new bank for slot `306720000` on MNB.

> don't understand your high level concern of how the polling negatively affects anything for this service itself being able to keep up, or just its normal operation

The concern is in regards to the overhead with how we are receiving messages from the channel. The comment is maybe a little less relevant with how your implementation has changed, but I'll still explain given how much we use crossbeam.

The spin-looping, yielding, and eventually getting blocked until woken up by sender all have cost. Again, if the queue is infinite in the eyes of this service (ie, it'll never be empty), then we'll never get to the spin-looping, yielding, and blocking so it is kind of a moot point. But, if we are able to ever clear the channel and have some "time-to-breath", we might actually hit [this section](https://github.com/crossbeam-rs/crossbeam/blob/45425b032b75d40c8f79be2133eb7d33aaa1d4e4/crossbeam-channel/src/flavors/list.rs#L469-L493). With 4 worker threads, maybe we would actually clear the queue (altho your response seemed to indicate maybe we still can't clear the queue at 40k TPS with 4 workers)

Thinking about MNB (where Brennan might have been going [here](https://github.com/anza-xyz/agave/pull/4032#discussion_r1880860082)), I'm led to believe that 1 worker is currently enough to keep up. So, the 3 additional workers would likely be hitting the spin/yield/blocked section frequently, which will "clog up" the system (extra system calls).

Definitely a tricky balance to optimize for current load but also have elasticity for higher load so we don't fall over.

> If the sleep optimization is about increasing CPU availability for other threads on other services, then it may be worth looking into but seems like a separate investigation and out of scope here.

I'd say the change I mentioned is more about making sure TransactionStatusService is a friendly service within the overall system. With more threads than cores, independent services can still impact each other in indirect ways

Reply:
Thanks. This is very helpful. Yeah that was a big miss on my part with the original implementation.